core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -690,12 +693,18 @@ public PlanWithProperties visitMergeWriter(MergeWriterNode node, PreferredProper
 
         private PlanWithProperties getWriterPlanWithProperties(Optional<PartitioningScheme> partitioningScheme, PlanWithProperties newSource, TableWriterNode.WriterTarget writerTarget)
         {
+            // use maxWritersTasks to set PartitioningScheme.partitionCount field to limit number of tasks that will take part in executing writing stage
+            int maxWriterTasks = writerTarget.getMaxWriterTasks(plannerContext.getMetadata(), session).orElse(getMaxWriterTasksCount(session));
```

Comment:
you can move `maxWriterTasks` variable to `if (partitioningScheme.isEmpty())` branch